src/bramble/client/ProjectStats.js

```diff
@@ -0,0 +1,258 @@
+define([
+    // Change this to filer vs. filer.min if you need to debug Filer
+    "thirdparty/filer/dist/filer.min",
+],function(Filer){
+    "use strict";
+
+    var ProjectStats = {};
+    
+    var _fs = new Filer.FileSystem();
+    ProjectStats.getFileSystem = function(){
+        if(_fs){
+            return _fs;
+        }
+        return null;
+    };
+
+    var _shell = new _fs.Shell();
+    var _Path = Filer.Path;
+    var _root;
+
+    // keep track of project state
+    // host cache in the object should any other project states have to be added
+    var _cache = {};
+
+    // walk the whole file tree and count its size
+    ProjectStats.init = function(root, callback){
+        function addSize(path, next){
+            // If the current path is a directory, record 0 for size
+            if(path.endsWith("/")) {
+                return next();
+            }
+
+            _fs.stat(path,function(err, stats){
+                _cache[path] = stats.size;
+            });
+        
+            next();
+        }
+
+        // reinitialize project stats
+        _root = root;
+        _cache = {};
+
+        // walk the root
+        _shell.find(_root, { exec:addSize }, function(err){
+            if(err) {
+               return callback(err);
+            }
+        });
+
+        callback(null);
+    };
+
+    // original unlink
+    var _innerUnlink = _fs.unlink;
+    // overwrite unlink function to do the bookkeeping of project state and call original unlink.
+    _fs.unlink = function(pathname, callback){
+        // call original Filer.unlink
+        _innerUnlink.call(_fs, pathname, function(err){
+            // only update cache once original was successful
+            if(!err){
+                // stop keeping track of deleted file
+                delete _cache[pathname];
+            }
+            
+            // call callback regardless
+            callback(err);
+        });
+    };
+
+    // original writeFile
+    var _innerWriteFile = _fs.writeFile;
+    // overwrite original writeFile and add bookkeeping of project state and call original writeFile.
+    _fs.writeFile = function(filename, data, options, callback){
```

Comment:
Let's simplify what you're doing here a bit, and just normalize the args before we use them (untested code follows):

```js
_fs.writeFile = function(filename, data, options, callback){
    if(typeof options === "function") {
        callback = options;
        options = {};
    }

    var size = data.length;

    _innerWriteFile.call(_fs, filename, data, options, function(err){
        if(err) {
            return callback(err);
        }
        // overwrite or add a new size for a given filename
        _cache[filename] = size;
        callback();
    });
};
```